docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1091 @@
+
+```cs
+unsafe int UnmanagedPointersSum(Span<int> buffer)
+{
+    fixed (int* pBuffer = buffer)
+    {
+        int* pEnd = pBuffer + buffer.Length;
+        int* pOneVectorFromEnd = pEnd - Vector128<int>.Count;
+        int* pCurrent = pBuffer;
+
+        Vector128<int> sum = Vector128<int>.Zero;
+
+        while (pCurrent <= pOneVectorFromEnd)
+        {
+            sum += Vector128.Load(pCurrent);
+
+            pCurrent += Vector128<int>.Count;
+        }
+
+        int result = Vector128.Sum(sum);
+
+        while (pCurrent < pEnd)
+        {
+            result += *pCurrent;
+
+            pCurrent++;
+        }
+
+        return result;
+    }
+}
+```
+
+`LoadAligned` and `LoadAlignedNonTemporal` require the input to be aligned. Aligned reads and writes should be slightly faster but using them comes at a price of increased complexity.
+
+Currently .NET exposes only one API fo allocating unmanaged aligned memory: [NativeMemory.AlignedAlloc](https://learn.microsoft.com/dotnet/api/system.runtime.interopservices.nativememory.alignedalloc). In the future, we might provide [a dedicated API](https://github.com/dotnet/runtime/issues/27146) for allocating managed, aligned and hence pinned memory buffers.
+
+The alternative to creating aligned buffers (we don't always have the control over input) is to pin the buffer, find first aligned address, handle non-aligned elements, then start aligned loop and afterwards handle the remainder. Adding such complexity to our code is hardly ever worth it and needs to be proved with proper benchmarking on various hardware.
+
+The fourth method expects only a managed reference (`ref T source`). We don't need to pin the buffer (GC is tracking managed references and updates them if memory gets moved), but it still requires us to properly handle managed pointer arithmetic:
+
+```cs
+int ManagedReferencesSum(int[] buffer)
+{
+    Debug.Assert(Vector128.IsHardwareAccelerated && buffer.Length >= Vector128<int>.Count);
+
+    ref int current = ref MemoryMarshal.GetArrayDataReference(buffer);
+    ref int end = ref Unsafe.Add(ref current, buffer.Length);
+    ref int oneVectorAwayFromEnd = ref Unsafe.Subtract(ref end, Vector128<int>.Count);
+
+    Vector128<int> sum = Vector128<int>.Zero;
+
+    while (!Unsafe.IsAddressGreaterThan(ref current, ref oneVectorAwayFromEnd))
+    {
+        sum += Vector128.LoadUnsafe(ref current);
+
+        current = ref Unsafe.Add(ref current, Vector128<int>.Count);
+    }
+
+    int result = Vector128.Sum(sum);
+
+    while (Unsafe.IsAddressLessThan(ref current, ref end))
+    {
+        result += current;
+
+        current = ref Unsafe.Add(ref current, 1);
+    }
```

Comment:
nit: It is much preferred to use `LoadUnsafe(ref bufferStart, index)` over using `Unsafe.Add`

It is very easy to get `Unsafe.Add` wrong, particularly with moving the `ref` outside the GC tracked region for the buffer. We have written a few bugs in that space ourselves.

`LoadUnsafe(ref T address, nuint index)` pushes users towards writing code that is less prone to this general issue and is overall easier to adapt to from a simpler `for` loop that the scalar logic may be using.

Reply:
-- I see we touch on this below, I think it would be much clearer to cover the "appropriate" way to do it here first in the normal reading order and then to cover things like this in a separate "what not to do and why" section.

Reply:
> I see we touch on this below, I think it would be much clearer to cover the "appropriate" way to do it here first in the normal reading order and then to cover things like this in a separate "what not to do and why" section.

That is why in the paragraph above (`Loops`), where I need to load the values, I am using the recommended way (so it's the first thing the users see). In this paragraph, I start on purpose with the "wrong" way, just to point out the disadvantages, move on & repat and at the end I mention the recommended approach.